src/internal/storage/kv/kv.go

```diff
@@ -1,32 +1,102 @@
+// Package kv provides a Key-Value Store interface and a few implementations.
 package kv
 
 import (
+	"bytes"
 	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/stream"
 )
 
 // ValueCallback is the type of functions used to access values
+// It is never ok for the callback to retain the data.
+// If it were ok, then the callee would just return the data.
 type ValueCallback = func([]byte) error
 
-// GetPut supports the basic Get and Put operations
-type GetPut interface {
-	// Get looks up the value that corresponds to key and calls cb once if the key exists
-	// if the key does not exist then a pacherr.ErrNotExist is returned
-	Get(ctx context.Context, key []byte, cb ValueCallback) error
+type Getter interface {
+	// Get looks up the value that corresponds to key and write the value into buf.
+	// If buf is too small for the value then io.ErrShortBuffer is returned.
+	Get(ctx context.Context, key []byte, buf []byte) (int, error)
+}
+
+type Putter interface {
 	// Put creates an entry mapping key to value, overwriting any previous mapping.
 	Put(ctx context.Context, key, value []byte) error
 }
 
+type GetPut interface {
+	Getter
+	Putter
+}
+
+type Deleter interface {
+	// Delete removes the entry at key.
+	// If there is no entry Delete returns nil
+	Delete(ctx context.Context, key []byte) error
+}
+
+type KeyIterable interface {
+	// NewKeyIterator returns a new iterator which will cover span.
+	NewKeyIterator(span Span) stream.Iterator[[]byte]
+}
+
 // Store is a key-value store
 type Store interface {
-	GetPut
-	Delete(ctx context.Context, key []byte) error
+	Getter
+	Putter
+	Deleter
 	Exists(ctx context.Context, key []byte) (bool, error)
-	Walk(ctx context.Context, prefix []byte, cb func(key []byte) error) error
+
+	KeyIterable
+}
+
+// Span is a range of bytes from Begin inclusive to End exclusive
+// As a special case if End == nil, then the span has no upper bound.
+type Span struct {
+	Begin []byte
+	End   []byte
+}
+
+// Contains returns true if the Span contains k
+func (s Span) Contains(k []byte) bool {
+	if bytes.Compare(s.Begin, k) > 0 {
+		return false
+	}
+	if s.End != nil && bytes.Compare(s.End, k) <= 0 {
+		return false
+	}
+	return true
+}
+
+func SpanFromPrefix(prefix []byte) Span {
+	return Span{
+		Begin: prefix,
+		End:   PrefixEnd(prefix),
+	}
+}
+
+// PrefixEnd returns the key > all the keys with prefix p, but < any other key
+func PrefixEnd(prefix []byte) []byte {
```

Comment:
If you look at the `Span` type in the other file, it has `Begin` and `End` fields which define a range including the lower bound, but excluding the upper bound.  The first key when listing a prefix is just the prefix itself, so that's what you set `Begin` to.  The key just after the end of the prefix is what you set `End` to.  `PrefixEnd` gives you the `End` key for the prefix that you pass in.